 design/README.MD

```diff
@@ -0,0 +1,456 @@
+DafnyTypes --> ToNative --> NativeTypes
+```
+
+However there may be many to one relationships.
+Perhaps `string` and `char[]`.
+These two would convert to `ValidUTF8Bytes` or `seq<uint8>`.
+
+Also reference types may not be exactly equal.
+While a native array would convert to a `seq`
+when converting back the new array
+will have the same elements in the same order
+but may not be exactly the same reference object.
+We can refer to this as `SemanticallyEqual`.
+
+The spirit of this can be expressed
+```dafny
+lemma ConversionsCorrectlyInvert<D, N>(ToDafny<D,N>, ToNative<N, D>, SemanticallyEqual)
+  ensures forall d, n ::
+    SemanticallyEqual(ToNative(d), n) <==> SemanticallyEqual(ToDafny(n), d)
+```
+
+Ideally these conversions would be written in Dafny
+so such properties could be proven.
+Either way we need to test
+that these things are true
+and that the correct duals are connected together.
+
+It is much easier if we can write our tests in Dafny.
+Then we don't need to write new tests for each runtime.
+This is where [wrapping a library](#wrapped-library) comes in.
+
+We want the test in Dafny and we want to test the conversions.
+by building a project we can control the Dafny implementation.
+By wrapping this project we can call it from Dafny
+and pass through the type conversions.
+This sandwiches the conversions between two functions we control.
+Namely, the implementation and the test.
+Many of these are already written in the TestModels directory.
+
+For the simple types: string, number, et al
+the implementation can not just echo back values.
+This is because it is possible that the conversion function
+do not produce rational input in the middle.
+This is an edge case, but it is easy enough to verify.
+
+Each simple type has smithy operations that:
+* returns back what is given.
+* only succeed if passed a specific value.
+* will only return a constant value
+
+We have one last subtlety to consider.
+Since the tests and implementation are both in Dafny
+we can clearly see the Dafny representation is reasonable.
+But this does not tell us that the native representation is reasonable.
+
+Consider a string.
+This string is represented under the hood as bytes
+and these bytes have an encoding.
+In `smithy-dafny` we represent this string as UTF8 bytes.
+If the conversion was to simply change the encoding to USC2
+then all our tests would pass.
+But looking at the native string it would be drastically different.
+
+This is a degenerate example but should illustrate the point.
+At this time we do not have any such degenerate tests written.
+This is because maintaining such coverage across multiple runtime
+is exceedingly difficult.
+It might be that a specification of such tests in conjunction with duvet
+might be a possible solution.
+
+## Project Management
+
+Neither Dafny nor `smithy-dafny` provide project management features.
+They are not terribly opinionated
+about how native project should be set up.
```

Comment:
Really? :) Seems to me like the shared makefile is very opinionated about how everything is laid out.

I think this is just another issue with the ambiguity of `smithy-dafny` though - the repo on the whole is opinionated because it includes the shared makefile, but you're trying to say the code generation tools don't care (which I think needs to change, since we should be generating more than just the code from the Smithy models)